File: 240419/황금비율 토스트/golden-toast.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(); // 식빵 개수
        int m = sc.nextInt(); // 암호문의 개수(커맨드 개수)

        char[] pw = sc.next().toCharArray();
        LinkedList<Character> ll = new LinkedList<Character>();
        for(int i = 0; i < n; i++){
            ll.addLast(pw[i]);
        }

        ListIterator<Character> iter = ll.listIterator(ll.size());

        for(int i = 0; i < m; i++){
            char[] cmd = sc.next().toCharArray();
            switch(cmd[0]){
                case 'L':
                    if(iter.hasPrevious()){
                        iter.previous();
                    }
                            // System.out.println(ll);

                    break;
                
                case 'R':
                    if(iter.hasNext()){
                        iter.next();
                    }
                            // System.out.println(ll);

                    break;

                case 'P':
                    char[] x = sc.next().toCharArray();
                    iter.add(x[0]);
                    // iter.next();
                            // System.out.println(ll);

                    break;

                case 'D':
                    if(iter.hasNext()){
                        iter.next();
                        iter.remove();
                    }
                    
                    // System.out.println(ll);

                    break;
                default:
                    break;
            }
        }

        for(char item : ll){
            System.out.print(item);
        }
    }
}

File: 240419/황금비율 토스트/golden-toast.kt
import java.util.*

fun main() {
    val sc = Scanner(System.`in`)
    val n = sc.nextInt() // bread count
    val m = sc.nextInt() // cmd count

    val ll = LinkedList<Char>()
    val pw = sc.next()

    for (i in 0 until n) {
        ll.add(pw[i])
    }

    // println(ll)

    val iter = ll.listIterator(ll.size)

    for(i in 0 until m){
        val cmd = sc.next()[0]
        when(cmd){
            'L' -> {
                if(iter.hasPrevious()){
                    iter.previous()
                }
            }
            'R' -> {
                if(iter.hasNext()){
                    iter.next()
                }
            }
            'D' -> {
                if(iter.hasNext()){
                    iter.next()
                    iter.remove()
                }
            }
            'P' -> {
                iter.add(sc.next()[0])
            }
        }
    }
    
    for(item in ll){
        print(item)
    }


}